test/e2e/workloadrebalancer_test.go

```diff
@@ -0,0 +1,180 @@
+		framework.CreateClusterPropagationPolicy(karmadaClient, policy)
+		framework.CreateDeployment(kubeClient, deploy)
+		framework.CreateDeployment(kubeClient, newAddedDeploy)
+		framework.CreateClusterRole(kubeClient, clusterrole)
+
+		ginkgo.DeferCleanup(func() {
+			framework.RemoveDeployment(kubeClient, deploy.Namespace, deploy.Name)
+			framework.RemoveDeployment(kubeClient, newAddedDeploy.Namespace, newAddedDeploy.Name)
+			framework.RemoveClusterRole(kubeClient, clusterrole.Name)
+			framework.RemoveClusterPropagationPolicy(karmadaClient, policy.Name)
+		})
+	})
+
+	var checkRescheduleResult = func(deployPossibleClusters [][]string, expectedWorkloads []appsv1alpha1.ObservedWorkload) {
+		// 1. check rebalancer status: match to `expectedWorkloads`.
+		framework.WaitRebalancerObservedWorkloads(karmadaClient, rebalancerName, expectedWorkloads)
+		// 2. check deploy: referenced binding's `spec.rescheduleTriggeredAt` and `status.lastScheduledTime` should be
+		// updated + actual propagation of replicas should change.
+		framework.WaitResourceBindingFitWith(karmadaClient, namespace, deployBindingName, func(rb *workv1alpha2.ResourceBinding) bool {
+			return bindingHasRescheduled(rb.Spec, rb.Status, rebalancer.CreationTimestamp)
+		})
+		framework.AssertBindingScheduledClusters(karmadaClient, namespace, deployBindingName, deployPossibleClusters)
+		// 3. check clusterrole: referenced binding's `spec.rescheduleTriggeredAt` and `status.lastScheduledTime` should be updated.
+		framework.WaitClusterResourceBindingFitWith(karmadaClient, clusterroleBindingName, func(crb *workv1alpha2.ClusterResourceBinding) bool {
+			return bindingHasRescheduled(crb.Spec, crb.Status, rebalancer.CreationTimestamp)
+		})
+	}
+
+	// 1. static weight scheduling
+	ginkgo.Context("static weight schedule type", func() {
+		ginkgo.BeforeEach(func() {
+			policy.Spec.Placement.ReplicaScheduling = helper.NewStaticWeightPolicyStrategy(targetClusters, []int64{2, 1})
+		})
+
+		ginkgo.It("reschedule when policy is static weight schedule type", func() {
+			ginkgo.By("step1: check first schedule result", func() {
+				// after first schedule, deployment is assigned as 2:1 in target clusters and clusterrole propagated to each cluster.
+				framework.AssertBindingScheduledClusters(karmadaClient, namespace, deployBindingName, [][]string{targetClusters})
+				framework.WaitClusterRolePresentOnClustersFitWith(targetClusters, clusterroleName, func(_ *rbacv1.ClusterRole) bool { return true })
+			})
+
+			ginkgo.By("step2: manually modify the schedule result", func() {
+				// directly modify deployment schedule result to mock failover when cluster failure and recovery.
+				// clusterrole not support this way, so skip mock clusterrole failover.
+				modifiedClusters := []workv1alpha2.TargetCluster{{Name: targetClusters[0], Replicas: *deploy.Spec.Replicas}}
+				framework.UpdateResourceBindingTargetClusters(karmadaClient, namespace, deployBindingName, modifiedClusters)
```

Comment:
1. I doubt we can `mock` failover like this. Feels a little bit tricky in E2E tests.
2. Do you have a test for the failover feature? How does it trigger the failover?

Reply:
> Do you have a test for the failover feature? How does it trigger the failover?

**Very complicated and will make my e2e case very bloated. In this case, I just want to test reschedule, failover is not the focus of my test.**

When I trigger failover in my local environment, I should do like this:

1. modify launch params of `controller-manager` to save failover waiting time from several minutes to several seconds.

```shell
kubectl --context karmada-host patch deploy karmada-controller-manager -n karmada-system --type='json' -p '[
   {"op": "replace", "path": "/spec/template/spec/containers/0/command/3", "value": "--cluster-status-update-frequency=3s"},
   {"op": "replace", "path": "/spec/template/spec/containers/0/command/5", "value": "--failover-eviction-timeout=3s"},
   {"op": "add", "path": "/spec/template/spec/containers/0/command/6", "value": "--cluster-failure-threshold=3s"},
   {"op": "add", "path": "/spec/template/spec/containers/0/command/7", "value": "--cluster-success-threshold=3s"}]'
```

2. modify `clusterTolerations` of policy to save failover toleration time.

```yaml
    clusterTolerations:
      - effect: NoExecute
        key: cluster.karmada.io/not-ready
        operator: Exists
        tolerationSeconds: 0
      - effect: NoExecute
        key: cluster.karmada.io/unreachable
        operator: Exists
        tolerationSeconds: 0
```

3. modify cluster apiEndpoint.

```shell
kubectl --context karmada-apiserver patch cluster member2 --type='json' -p '[{"op": "replace", "path": "/spec/apiEndpoint", "value": "https://172.18.0.4:6444"}]'
```

4. waiting for cluster failover.

5. modify cluster apiEndpoint back and wait cluster ready again, and then create a `WorkloadRebalancer` to trigger a reschedule.

Reply:
This will also affect other parallel cases.